src/System.IO.FileSystem/ref/System.IO.FileSystem.cs

```diff
@@ -212,3 +224,75 @@ public enum SearchOption
         TopDirectoryOnly = 0,
     }
 }
+namespace System.IO.Enumeration
+{
+    public enum MatchType
+    {
+        Simple,
+        Dos
+    }
+    public class EnumerationOptions
+    {
+        public EnumerationOptions() { }
+
+        public bool RecurseSubdirectories { get { throw null; } set { } }
+        public bool IgnoreInaccessible { get { throw null; } set { } }
+        public int BufferSize { get { throw null; } set { } }
+        public FileAttributes AttributesToSkip { get { throw null; } set { } }
+        public MatchType MatchType { get { throw null; } set { } }
+        public bool ReturnSpecialDirectories { get { throw null; } set { } }
+    }
+    public ref struct FileSystemEntry
+    {
+        public ReadOnlySpan<char> Directory { get { throw null; } }
+        public string RootDirectory { get { throw null; } }
+        public string OriginalRootDirectory { get { throw null; } }
+        public ReadOnlySpan<char> FileName { get { throw null; } }
+        public FileAttributes Attributes { get { throw null; } }
+        public long Length { get { throw null; } }
+        public DateTimeOffset CreationTimeUtc { get { throw null; } }
+        public DateTimeOffset LastAccessTimeUtc { get { throw null; } }
+        public DateTimeOffset LastWriteTimeUtc { get { throw null; } }
+        public bool IsDirectory { get { throw null; } }
+        public FileSystemInfo ToFileSystemInfo() { throw null; }
+        public string ToSpecifiedFullPath() { throw null; }
+        public string ToFullPath() { throw null; }
+    }
+    public abstract class FileSystemEnumerator<TResult> : Runtime.ConstrainedExecution.CriticalFinalizerObject, Collections.Generic.IEnumerator<TResult>
```

Comment:
I missed the API review while I was out... why does this need to be public?  It's not just an implementation detail of the enumerable?

Reply:
There are two options for extending, derive the enumerator directly, or use our enumerable wrapper and delegates. The enumerable should cover most needs and keep the type count contained. The enumerator will have some of the much narrower use cases, such as wanting to know when a directory is finished, handle raw error codes, etc..